buildAndReleaseTask/serviceEndpoint.ts

```diff
@@ -47,4 +56,84 @@ export function getServiceEndpoint(
 
     return endpoint;
 }
+
+/**
+ * Tries to get the endpoint details for a workload identity federation
+ * service connection as per V2 of the Azure CLI task
+ * https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/AzureCLIV2/azureclitask.ts#L143-L152
+ */
+async function tryGetOidcServiceEndpoint(
+    connectedServiceName: string
+): Promise<IServiceEndpoint | undefined> {
+    const authScheme = tl.getEndpointAuthorizationScheme(
+        connectedServiceName,
+        true
+    );
+    if (authScheme?.toLowerCase() !== "workloadidentityfederation") {
+        return undefined;
+    }
+
+    const federatedToken = await tryGetIdToken(connectedServiceName);
+
+    if (!federatedToken) {
+        return undefined;
+    }
+
+    tl.setSecret(federatedToken);
+
+    const endpoint = {
+        clientId: tl.getEndpointAuthorizationParameter(
+            connectedServiceName,
+            "serviceprincipalid",
+            false
+        ),
+        oidcToken: await tryGetIdToken(connectedServiceName),
+        // It is entirely possible subscriptionId is not present, so setting subscription as optional stops an unintended failure.
+        // eg- a ManagementGroup scoped SP has access to multiple subscriptions; the subscription that is the target of a pulumi up will be enforced in the pulumi program config/env.
+        subscriptionId: tl.getEndpointDataParameter(
+            connectedServiceName,
+            "subscriptionid",
+            true
+        ),
+        tenantId: tl.getEndpointAuthorizationParameter(
+            connectedServiceName,
+            "tenantid",
+            false
+        ),
+    } as IServiceEndpoint;
+
+    return endpoint;
+}
+
+/**
+ * Tries to get the ID token as per V2 of the Azure CLI task
+ * https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/AzureCLIV2/azureclitask.ts#L251-L268
+ */
+async function tryGetIdToken(
+    connectedService: string
+): Promise<string | undefined | null> {
+    const jobId = tl.getVariable("System.JobId")!;
+    const planId = tl.getVariable("System.PlanId")!;
+    const projectId = tl.getVariable("System.TeamProjectId")!;
+    const hub = tl.getVariable("System.HostType")!;
+    const uri = tl.getVariable("System.CollectionUri")!;
+    const token = getSystemAccessToken();
+
+    const authHandler = getHandlerFromToken(token);
+    const connection = new WebApi(uri, authHandler);
+    const api: ITaskApi = await connection.getTaskApi();
+    const response = await api.createOidcToken(
+        {},
+        projectId,
+        hub,
+        planId,
+        jobId,
+        connectedService
+    );
+    if (response == null) {
+        return null;
```

Comment:
It looks like in this case `tryGetOidcServiceEndpoint` returns an endpoint but with a null token, which is then returned in `getServiceEndpoint` but will fail later. Maybe we should check this case and log a clear error message earlier?

Reply:
If the returned value is falsy in `getServiceEndpoint` at line [27](https://github.com/Isenr/pulumi-az-pipelines-task/blob/ff9f7bd095ef971cb3bd54b08d7f938024cf94bd/buildAndReleaseTask/serviceEndpoint.ts#L27) the function won't return and will execute the previous behaviour.